Add libavrocpp/1.10.1

plopp opened the pull request:
Specify library name and version:  **libavrocpp/1.10.1**

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

prince-chrismc commented on recipes/libavrocpp/1.10.1/conanfile.py:
```diff
@@ -0,0 +1,100 @@
+class LibavrocppConan(ConanFile):
+    name = "libavrocpp"
+    version = "1.10.1"
+    license = "Apache License 2.0"
```
We only used the SPDX short identifiers

```suggestion
    license = "Apache-2.0"
```

prince-chrismc commented on recipes/libavrocpp/1.10.1/conanfile.py:
```diff
@@ -0,0 +1,100 @@
+
+class LibavrocppConan(ConanFile):
+    name = "libavrocpp"
+    version = "1.10.1"
```
```suggestion
```

You can set this via the command line `conan create . 1.10.1@` 👍

prince-chrismc commented on recipes/libavrocpp/1.10.1/conanfile.py:
```diff
@@ -0,0 +1,100 @@
+            "apple-clang": "5",
+        }
+
+        if compiler in minimal_version and compiler_version < minimal_version[compiler]:
```
We like to print a warning when the the compiler is not known.

conan-io/conan#8002 has a nice boilerplate you can copy...

If you require the granularity for the compiler version you need to use the `lazy_semver_lt` from Space https://github.com/conan-io/conan-center-index/commit/325c293a35ec94477c9e5197190190d2d00add97

None of the compilers excluded are support on CCI this is not required and can be removed.

prince-chrismc commented on recipes/libavrocpp/1.10.1/conanfile.py:
```diff
@@ -0,0 +1,100 @@
+    def source(self):
+        tools.get(**self.conan_data["sources"][self.version])
+        os.rename("avro-release-" + self.version, "source_subfolder")
+        tools.replace_in_file(
```
replace in file and patch MUST be applied during the build method

prince-chrismc commented on recipes/libavrocpp/config.yml:
```diff
@@ -0,0 +1,3 @@
+versions:
+  1.10.1:
+    folder: 1.10.1
```
I can not see this project change so drastically that a new recipe is required for each patch release 

Please more the recipe to an `all` folder

prince-chrismc commented on recipes/libavrocpp/1.10.1/conanfile.py:
```diff
@@ -0,0 +1,100 @@
+               conan_basic_setup()""",
+        )
+
+        if self.options.with_snappy:
+            tools.replace_in_file(
+                os.path.join(self._source_subfolder, "CMakeLists.txt"),
+                "find_package(Snappy)",
+                "",
+            )
```
If upstream does not support this as an option I suspect it breaks the #3903 rules that are pending. I suspect this may need to be removed.

plopp replied:
Upstream supports it 100%, but the included `FindSnappy.cmake` does not play well with the injection of `conanbuildinfo.cmake`. When cmake calls find_package(Snappy) it seems to mess up the variables set by conan in the previous step. I guess I'm not the first one to have this problem, so there is probably a better solution. #3903 was unknown to me, but I don't see this patch breaking that, does it?

prince-chrismc commented on recipes/libavrocpp/1.10.1/conanfile.py:
```diff
@@ -0,0 +1,100 @@
+            del self.options.fPIC
+
+        compiler = str(self.settings.compiler)
+        compiler_version = Version(self.settings.compiler.version.value)
```
The value is not required 
```suggestion
        compiler_version = Version(self.settings.compiler.version)
```

prince-chrismc commented on recipes/libavrocpp/1.10.1/conanfile.py:
```diff
@@ -0,0 +1,100 @@
+    def package(self):
+        self.copy("LICENSE", dst="licenses", src=self._source_subfolder)
+        self.copy("NOTICE", dst="licenses", src=self._source_subfolder)
+        cmake = CMake(self)
```
We like to encourage cache the CMake as a member to prevent the slow configuration from taking place

prince-chrismc reviewed (changes requested):
Thank you for this contribution 

One think which would be appreciated would be to follow this order 
https://github.com/conan-io/conan-center-index/blob/master/docs/reviewing.md#order-of-methods-and-attributes

There's quite a few changes required so hopefully its reasonable request

If you need any help feel free to ask!

I

prince-chrismc commented on recipes/libavrocpp/1.10.1/conanfile.py:
```diff
@@ -0,0 +1,100 @@
+        self.copy("LICENSE", dst="licenses", src=self._source_subfolder)
+        self.copy("NOTICE", dst="licenses", src=self._source_subfolder)
+        cmake = CMake(self)
+        cmake.configure(source_folder=self._source_subfolder)
```
This by passes the cmake_wrapper you added (it's requires to push all the correct settings for CCI) typically you only need to set the build folder

plopp replied:
Alright, thanks, I'm not entirely certain I understand the full extent of it, but I called `cmake = self._configure_cmake()` instead and then just called `cmake.install()`

plopp commented:
Thanks @prince-chrismc! There are two PR:s up, one for 1.10.1 and one for 1.8.2. Both of them were in one PR from the beginning. I split them up and reworked the 1.8.2 first, to deal with this one later. I will apply your comments to the other one as well: #4391

prince-chrismc commented:
Why are you adding two versions? unless there's a specific need for the older I'd strongly suggest you only keep this one

prince-chrismc commented:
You are probable not correct;y following the CMake exported definitions, https://github.com/apache/avro/blob/7d1e63b219e6d0778bc57195152477adee97fcab/lang/c%2B%2B/api/Config.hh#L31
Seems like it think it's supposed to be shared 

make sure `AVRO_DYN_LINK` is not defined

plopp commented:
Thanks! There was a specific need for the old one because of another dependency. I will revisit that. The 1.8.2 is old enough to warrant an upgrade across the board on my end. Thank you for the review so far!

prince-chrismc commented on recipes/libavrocpp/all/conanfile.py:
```diff
@@ -67,34 +52,42 @@ def source(self):
-        cmake = CMake(self)
-        cmake.configure(source_folder=self._source_subfolder)
-        cmake.build()
+        if self.options.shared:
+            self.cpp_info.components["avrocpp"].libs = ["avrocpp"]
+            self.cpp_info.components["avrocpp"].requires = ["boost::boost"]
+        else:
+            self.cpp_info.components["avrocpp_s"].libs = ["avrocpp_s"]
+            self.cpp_info.components["avrocpp_s"].requires = ["boost::boost"]
```
```suggestion
        self.cpp_info.components["avrocpp"].libs = ["avrocpp" if self.options.shared else "avrocpp_s"]
        self.cpp_info.components["avrocpp"].requires = ["boost::boost"]
```

prince-chrismc commented on recipes/libavrocpp/all/conanfile.py:
```diff
@@ -67,34 +52,42 @@ def source(self):
         self.requires("boost/[>=1.38.0]")
+
+    def build_requirements(self):
+        self.build_requires("boost/[>=1.38.0]")
```
You should only add boost once, we do do not use version ranges in CCI (to ensure reproducible builds)

plopp commented:
Thanks for the frequent reviews @prince-chrismc!

plopp commented:
So the VS builds are a real headache. Summary of errors for each CCI-config:

- VS16 Debug Static MTd: 
```
error LNK2019: unresolved external symbol "__declspec(dllimport) public: class boost::program_options::options_description_easy_init & __cdecl boost::program_options::options_description_easy_init::operator()(char const *,char const *)"
```
- VS14 Debug Static MDd: `error C3861: 'back_inserter': identifier not found`
- VS14 Debug Static MTd: `error C3861: 'back_inserter': identifier not found`
- VS14 Release Static MT: `error C3861: 'back_inserter': identifier not found`
- VS14 Release Static MD: `error C3861: 'back_inserter': identifier not found`
- VS15 Release Static MT, VS15 Release Static MTd, VS15 Release Static MD: 
```
avrogencpp.obj : error LNK2019: unresolved external symbol "__declspec(dllimport) public: virtual __cdecl boost::program_options::value_semantic_codecvt_helper<char>::~value_semantic_codecvt_helper<char>(void)" (__imp_??1?$value_semantic_codecvt_helper@D@program_options@boost@@UEAA@XZ)
```
I still have a TODO to verify that `AVRO_DYN_LINK` is properly set for shared but remains unset for static builds. Apart from that I'm not sure how to proceed. I will have to do a deeper investigation. @prince-chrismc

prince-chrismc commented on recipes/libavrocpp/all/conanfile.py:
```diff
@@ -0,0 +1,90 @@
+            tools.check_min_cppstd(self, "11")
+
+    def requirements(self):
+        self.requires("boost/[1.75.0]")
```
```suggestion
        self.requires("boost/1.75.0")
```

plopp replied:
oh, typo.

prince-chrismc commented on recipes/libavrocpp/all/conanfile.py:
```diff
@@ -0,0 +1,90 @@
+    def requirements(self):
+        self.requires("boost/[1.75.0]")
+
+    def build_requirements(self):
+        if self.options.with_snappy:
+            self.build_requires("snappy/1.1.8")
```
Conan does not support private dependencies as of yet, I suspect this should be a requirement

plopp replied:
Ok, I wanted to only depend on snappy during build, and not when linking against the avro shared library.

prince-chrismc replied:
Sadly this would not work for consumers using statis builds

plopp commented:
I just noticed your new comments just after hitting send on my last message there. Will have a look at those as well.

prince-chrismc commented on recipes/libavrocpp/all/conanfile.py:
```diff
@@ -0,0 +1,90 @@
+        cmake.install()
+
+    def package_info(self):
+        self.cpp_info.components["avrocpp"].libs = [
+            "avrocpp" if self.options.shared else "avrocpp_s"
+        ]
+        self.cpp_info.components["avrocpp"].requires = ["boost::boost"]
```
Sorry I broke the components in my last review 🙊 

```suggestion
        target = "avrocpp" if self.options.shared else "avrocpp_s"
        self.cpp_info.components[target].libs = [target]
        self.cpp_info.components[target].requires = ["boost::boost"]
```

plopp replied:
Ok, but from what I could see it was working anyway. The test compiled fine for me only using the avrocpp target. I even checked the dependencies with `ldd`

prince-chrismc replied:
I understand but we try to correct model the CMake targets from upstream

https://github.com/apache/avro/blob/6d77f5048dcc3b6f0924c86b740a2c90a8354fa8/lang/c%2B%2B/CMakeLists.txt#L193

Which is why I made the suggestion above.... this way the with default name the components will be generated correctly

plopp replied:
Got it. Thanks!

prince-chrismc commented on recipes/libavrocpp/all/test_package/CMakeLists.txt:
```diff
@@ -0,0 +1,12 @@
+
+set_property(TARGET ${PROJECT_NAME} PROPERTY CXX_STANDARD 11)
+
+target_link_libraries(${PROJECT_NAME} libavrocpp::avrocpp)
```
```suggestion
if(AVROCPP_SHARED)
    target_link_libraries(${PROJECT_NAME} libavrocpp::avrocpp)	
else()	
    target_link_libraries(${PROJECT_NAME} libavrocpp::avrocpp_s)	
endif()
```

prince-chrismc commented:
> Thanks for the frequent reviews @prince-chrismc!

❤️ Thank you for the contribution 🏆 

> I just noticed your new comments just after hitting send on my last message there. Will have a look at those as well.

I think we have the same routine 🤣

plopp commented on recipes/libavrocpp/all/conanfile.py:
```diff
@@ -0,0 +1,93 @@
+        self._cmake.configure(source_folder=self._source_subfolder)
+        return self._cmake
+
+    def configure(self):
```
Oh, I thought I fixed that. Are you referring to the extra `_configure_cmake()` method? The other ones are in the correct order, otherwise I have overlooked something. I fixed that in a previous commit.

prince-chrismc replied:
Having problems with the UI this morning....

I forgot to submit this last night so it might be stale

prince-chrismc commented:
The missing symbols are all for the static runtimes (which are probably not support) the missing names might be this https://github.com/erengy/anitomy/issues/2#issuecomment-124042738

plopp commented on recipes/libavrocpp/all/conanfile.py:
```diff
@@ -0,0 +1,93 @@
+               conan_basic_setup()""",
+        )
+
+        if self.options.with_snappy:
+            tools.replace_in_file(
+                os.path.join(self._source_subfolder, "CMakeLists.txt"),
+                "find_package(Snappy)",
+                "",
+            )
```
I tried setting the variable to an empty string in `_configure_cmake()` when *not* using snappy:
```
if not self.options.with_snappy:
    self._cmake.definitions["SNAPPY_ROOT_DIR"] = ""
```
But that failed:
```
-- Could NOT find Snappy (missing: SNAPPY_LIBRARIES SNAPPY_INCLUDE_DIR)
Disabled snappy codec. libsnappy not found.
-- Configuring done
CMake Error: The following variables are used in this project, but they are set to NOTFOUND.
Please set them or make sure they are set and tested correctly in the CMake files:
/home/user/.conan/data/libavrocpp/1.10.1/_/_/build/a285fc00ea829acfc7280102b9b32a8043fbd51c/source_subfolder/lang/c++/SNAPPY_INCLUDE_DIR
   used as include directory in directory /home/user/.conan/data/libavrocpp/1.10.1/_/_/build/a285fc00ea829acfc7280102b9b32a8043fbd51c/source_subfolder/lang/c++
   used as include directory in directory /home/user/.conan/data/libavrocpp/1.10.1/_/_/build/a285fc00ea829acfc7280102b9b32a8043fbd51c/source_subfolder/lang/c++
   ```
   I will leave the removal of the builtin call to `find_package(snappy)` in there meanwhile, and see if I can find a better way later.

plopp commented:
Hmm, I can't see the actual error, I only see `This commit cannot be built`. Clicking `Details` gives me `403 Forbidden`. @prince-chrismc

plopp commented:
Ok, i fixed the comments, I am somewhat at a loss here because of the linking problems for Windows. What do you suggest now @prince-chrismc?

prince-chrismc commented:
@plopp Please consider https://github.com/plopp/conan-center-index/pull/1

It was a lot of hard work 😅 Windows static release + windows shared debug work after that

prince-chrismc commented:
11hr build time... 😞 CCI bug?

Please close, wait 10s and reopen the PR

prince-chrismc commented:
😡 https://c3i.jfrog.io/c3i/misc/logs/pr/4393/13/libavrocpp/1.10.1/b58a5b725d6f9d8c0c16afcad85e4b077fef00e1/create_stderr.txt

It's listed online https://conan.io/center/zlib?os=Windows&tab=configuration

```
ERROR: Missing binary: zlib/1.2.11:b2b69efcc79a149e0bc765fb12ba0278d52fc260
zlib/1.2.11: WARN: Can't find a 'zlib/1.2.11' package for the specified settings, options and dependencies:
- Settings: arch=x86_64, build_type=Debug, compiler=Visual Studio, compiler.runtime=MTd, compiler.version=16, os=Windows
- Options: minizip=False, shared=False
- Dependencies: 
- Requirements: 
- Package ID: b2b69efcc79a149e0bc765fb12ba0278d52fc260

ERROR: Missing prebuilt package for 'zlib/1.2.11'
Try to build from sources with '--build=zlib'
Use 'conan search <reference> --table table.html'
Or read 'http://docs.conan.io/en/latest/faq/troubleshooting.html#error-missing-prebuilt-package'
```

prince-chrismc commented:
@ericLemanissier can you please add this too #4221?

plopp commented:
@prince-chrismc I'm on a slipper slope here, but I noticed when I clicked that specific package ID (b2b69efcc79a149e0bc765fb12ba0278d52fc260) on conancenter it showed this conan info where **shared=False** but the failing job was using **shared=True**:

```
[settings]
    arch=x86_64
    build_type=Debug
    compiler=Visual Studio
    compiler.runtime=MTd
    compiler.version=16
    os=Windows

[requires]


[options]
    minizip=False
    shared=False

[full_settings]
    arch=x86_64
    build_type=Debug
    compiler=Visual Studio
    compiler.runtime=MTd
    compiler.version=16
    os=Windows

[full_requires]


[full_options]
    minizip=False
    shared=False

[recipe_hash]
    25dad5342ca50f8d7b9e30f669c7ad37

[env]

```
The conan info path was: `_/zlib/1.2.11/_/25dad5342ca50f8d7b9e30f669c7ad37/b2b69efcc79a149e0bc765fb12ba0278d52fc260/50f62da3fe8881bf6cb23f20fdd29de8`

plopp commented:
Nevermind the comment I deleted, I overlooked something.

plopp commented:
I restarted CCI based on this comment from #4221:

> EDIT: removing this one, because the binaries were probably being regerated for [a408727](https://github.com/conan-io/conan-center-index/commit/a40872784b83cfaf67efb65f501875e08b1427de) at this moment: [#4393 (comment)](https://github.com/conan-io/conan-center-index/pull/4393#issuecomment-772562987): missing zlib binary (also on [#4412 (comment)](https://github.com/conan-io/conan-center-index/pull/4412#issuecomment-772559810) and [#4472 (comment)](https://github.com/conan-io/conan-center-index/pull/4472#issuecomment-772561325))

prince-chrismc reviewed (dismissed):
LGTM 🚀

uilianries commented on recipes/libavrocpp/all/conanfile.py:
```diff
@@ -0,0 +1,89 @@
+        for patch in self.conan_data.get("patches", {}).get(self.version, []):
+            tools.patch(**patch)
+        tools.replace_in_file(
+            os.path.join(self._source_subfolder, "CMakeLists.txt"),
+            "project (Avro-cpp)",
+            """project (Avro-cpp)
+               include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
+               conan_basic_setup()"""
```
Try to use CMake wrapper instead. It's more compatible in case something change in the future. Also it's most popular pattern.

plopp replied:
I will give it a try, thanks!

prince-chrismc commented:
@plopp please close the other PR #4391 🙏 if you really need that version you can add it to this recipe 👍 after it's merged

uilianries reviewed (approved):
LGTM
